File: Pandacap.Data/FeedItem.cs
namespace Pandacap.Data
{
    public class FeedItem : IPost
    {
        public Guid Id { get; set; }

        public string? FeedTitle { get; set; }

        public string? FeedWebsiteUrl { get; set; }

        public string? FeedIconUrl { get; set; }

        public string? Title { get; set; }

        public string? Url { get; set; }

        public string? HtmlDescription { get; set; }

        public DateTimeOffset Timestamp { get; set; }

        private record Image(string Url) : IPostImage
        {
            string? IPostImage.ThumbnailUrl => Url;

            string? IPostImage.AltText => null;
        }

        public IEnumerable<IPostImage> Images
        {
            get
            {
                string work = HtmlDescription ?? "";
                while (true)
                {
                    int index = work.IndexOf("src=", StringComparison.InvariantCultureIgnoreCase);
                    if (index == -1)
                        break;

                    work = work[(index + 4)..];

                    char stringDelimeter = work.DefaultIfEmpty('\0').First();
                    if (stringDelimeter != '"' && stringDelimeter != '\'')
                        continue;

                    work = work[1..];

                    int endIndex = work.IndexOf(stringDelimeter);
                    string url = work[..endIndex];
                    work = work[endIndex..];

                    yield return new Image(url);
                }
            }
        }

        string IPost.Id => $"{Id}";

        string? IPost.Username => FeedTitle;

        string? IPost.Usericon => FeedIconUrl;

        string IPost.DisplayTitle => Title ?? $"{Id}";

        DateTimeOffset IPost.Timestamp => Timestamp;

        string? IPost.LinkUrl => Url;
    }
}

File: Pandacap.Data/Feeds/FeedItem.fs
namespace Pandacap.Data

open System
open FSharp.Data

type FeedItem() =
    member val Id = Guid.Empty with get, set
    member val FeedTitle = nullString with get, set
    member val FeedWebsiteUrl = nullString with get, set
    member val FeedIconUrl = nullString with get, set
    member val Title = nullString with get, set
    member val Url = nullString with get, set
    member val HtmlDescription = nullString with get, set
    member val Timestamp = DateTimeOffset.MinValue with get, set

    interface IPost with
        member this.DisplayTitle = this.Title |> orString $"{this.Id}"
        member this.Id = $"{this.Id}"
        member this.Images =
            let html = HtmlDocument.Parse (this.HtmlDescription |> orString "")

            html.Descendants "img"
            |> Seq.choose (fun node -> node.TryGetAttribute "src")
            |> Seq.map (fun attr -> attr.Value())
            |> Seq.map (fun src -> {
                new IPostImage with
                    member _.AltText = null
                    member _.ThumbnailUrl = src
            })
        member this.LinkUrl = this.Url
        member this.Timestamp = this.Timestamp
        member this.Usericon = this.FeedIconUrl
        member this.Username = this.FeedTitle
